### PR TITLE
3485787 add message reordering patch

### PR DESCRIPTION
**Pull request by elopezanaya**
### Id of the task, bug, story or other reference.

internal task : 3485787 
+ bug : 3552041

### Description
_Include a description of the problem to be solved_

Use of a new tag value as part of the payload to override webchat:sequence-id using OriginalMessageId value.

## Solution Proposed
_Detail what is the solution proposed, including links to the design document if required or any other document required to support the solution_

Middleware to evaluate all incoming activity to ensure the proper sorting and display of the messages.

__additional to the original purpose, I'm including a change to handle the proper sorting of the messages in the transcript__

### Acceptance criteria
_Define what are the conditions to consider the PR has achieved the intended goal_

- C1&C2 Conversations honors the intended order of the messages based on the originalMessageId when is present
- PVA&C2 Conversations honors the intended order of the messages based on the originalMessageId when is present
- BYOB&C2 Conversations honors the intended order of the messages based on the originalMessageId when is present
- includes message system 
- middleware only overrides the value of webchat:sequence-id with originalMessageId when is present 

## Test cases and evidence
_Include what tests cases were considered, any evidence of testing for future references, to identify any corner cases, etc_

#### C1&C2

data-app-id="777dc572-a896-485e-8be9-e40cb68d35d4" 
data-org-id="5c724e8f-0a05-ee11-a66e-00224820c232" 
data-org-url="https://unq5c724e8f0a05ee11a66e00224820c-crm10.oc.crmlivetie.com"

image

image

#### C1 C2 with the script to simulate fast input

image
image

#### PVA - C2

data-app-id="dcb1efec-fa05-4b1e-8603-6680b0a0a5da" 
data-org-id="bb5ed8ab-baff-ed11-a66b-000d3a5bdef8" data-org-url="https://unqbb5ed8abbaffed11a66b000d3a5bd-crmtest.oc.crmlivetie.com"

trigger : scramble 

image

image

#### BYOB -C2

data-org-id="bb5ed8ab-baff-ed11-a66b-000d3a5bdef8" 
data-app-id="56ec6220-c6ea-4ec3-b5a5-9a5b138e7fd8" data-org-url="https://unqbb5ed8abbaffed11a66b000d3a5bd-crmtest.oc.crmlivetie.com" 

#### Conversation with org with no tag present in the payload

image

__there is no action for multiple messages, so the test is requesting multiple cards and corroborating order__

image

image

image

#### C1&C2 Transcripts display in proper order

image
image

#### BYOB transcript display in proper order
image

#### PVA transcript display in order

image

#### When OriginalMessageId is not present, sorting for chat and transcript is not guaranteed

image

### Sanity Tests
-   [ ] You have tested all changes in Popout mode
-   [ ] You have tested all changes in cross browsers i.e Edge, Chrome, Firefox, Safari and mobile devices(iOS and Android)
-   [ ] Your changes are included in the [CHANGELOG](../CHANGE_LOG.md)

## A11y
- [ ] You have run the [Automated Accessibility Check](https://accessibilityinsights.io/docs/en/windows/getstarted/automatedchecks) and have no reported issues

__*Please provide justification if any of the validations has been skipped.*__

### CODE REVIEW

File: CHANGE_LOG.md [hunk 1]
```diff
@@ -13,7 +13,7 @@ All notable changes to this project will be documented in this file.
 - Uptake [@microsoft/omnichannel-chat-sdk@1.4.5](https://www.npmjs.com/package/@microsoft/omnichannel-chat-sdk/v/1.4.5)
 - Added parity to NewMessage event payload, and renamed `HistoryMessageReceived` to `RehydrateMessageReceived` to avoid conflict with the same BroadcastEvent
 - Added Pull Request Template
-
+- Added new middleware (messageSequenceIdOverrideMiddleware) to ensure proper order of messages
```
Comment (charliewang95) [inline CHANGE_LOG.md:16]: Move this up to "Unreleased" section

File: chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/renderingmiddlewares/messageSequenceIdOverrideMiddleware.tsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+const isApplicable = (action: IWebChatAction): boolean => {
+
+    return action.type === WebChatActionType.DIRECT_LINE_INCOMING_ACTIVITY &&
+        isValidChannel(action) &&
+        lookupOriginalMessageId(action) !== undefined &&
+        isWebSequenceIdPresent(action);
+};
+
+const isValidChannel = (action: IWebChatAction): boolean => {
```
Comment (charliewang95) [inline chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/renderingmiddlewares/messageSequenceIdOverrideMiddleware.tsx:21]: I remember there are other middlewares using similar logics. Can we extract them into a common place, so that we can add unit test on them?

**Comment by kevinlin98z** [comment]
Please help test message ordering using this CC script snippet to send multiple messages fast as well:

```
let uid = Math.random() * 1000000000;
uid = Math.floor(uid) + 1000000000;

 

text = `${uid} cc message`;

 

function waitFor(ms) {
    return new Promise(resolve => {
        setTimeout(() => {
            resolve();
        }, ms)
    })
}

 

async function sendRepeatMessages() {
    for (let index = 0; index < 50; index++) {
        let testPayload = { type: 'WEB_CHAT/SEND_MESSAGE', payload: { text: text + " " + index} }

 

        Microsoft.OmniChannel.ConversationControl.ConversationManager.getInstance().getCurrentConversation().webChatStore.dispatch(testPayload);
        await waitFor(100)
    }
}

 

sendRepeatMessages()
```

File: chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/renderingmiddlewares/messageSequenceIdOverrideMiddleware.tsx (hunk 2 above)
Comment (elopezanaya) [reply chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/renderingmiddlewares/messageSequenceIdOverrideMiddleware.tsx:21]: the other middleware (createMessageTimeStampMiddleware) using it will be removed in a posterior task since is not in use

File: CHANGE_LOG.md [hunk 3]
```diff
@@ -5,6 +5,7 @@ All notable changes to this project will be documented in this file.
 # Chat Widget
 
 ## [Unreleased]
```
Comment (charliewang95) [inline CHANGE_LOG.md:7]: Missing an `###Added` section name above this item

File: chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/renderingmiddlewares/messageSequenceIdOverrideMiddleware.tsx [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    });
+};
+
+const overrideSequenceIdWithOriginalMessageId = (action: IWebChatAction): IWebChatAction => {
+
+    const originalMessageId = extractOriginalMessageId(action);
+
+    if (originalMessageId === undefined) return action;
+
+    Object.keys(action.payload.activity.channelData).forEach(function (key) {
```
Comment (charliewang95) [inline chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/renderingmiddlewares/messageSequenceIdOverrideMiddleware.tsx:44]: nit: I'd assign a separate variable to store `action.payload.activity.channelData` in this method, for readability

File: docs/pull_request_template.md [hunk 5]
```diff
@@ -1,25 +0,0 @@
-## **Thank you for your contribution. Before submitting this PR, please include:**
```
Comment (charliewang95) [inline docs/pull_request_template.md:1]: Why is this removed?

File: chat-widget/src/plugins/createChatTranscript.ts [hunk 6]
```diff
@@ -210,9 +222,22 @@ class TranscriptHTMLBuilder {
...
-                                timestamp: created
+                                timestamp: created,
+                                channelData: { 
+                                    "webchat:sequence-id": webchatSequenceId
+                                },
                             };
                         }
+                        
+                        static convertStringValueToInt(value) {
+
```
Comment (charliewang95) [inline chat-widget/src/plugins/createChatTranscript.ts:233]: nit: empty line

File: chat-widget/src/plugins/createChatTranscript.ts [hunk 7]
```diff
@@ -186,7 +192,10 @@ class TranscriptHTMLBuilder {
                                 return {
                                     ...activity,
                                     text,
-                                    timestamp: created
+                                    timestamp: created,
+                                    channelData: { 
+                                        "webchat:sequence-id": webchatSequenceId
+                                    },
```
Comment (charliewang95) [inline chat-widget/src/plugins/createChatTranscript.ts:198]: nit: trailing comma

File: chat-widget/src/plugins/createChatTranscript.ts [hunk 8]
```diff
@@ -199,7 +208,10 @@ class TranscriptHTMLBuilder {
                                         return {
                                             ...activity,
                                             ...partialActivity,
-                                            timestamp: created
+                                            timestamp: created,
+                                            channelData: { 
+                                                "webchat:sequence-id": webchatSequenceId
+                                            },
```
Comment (charliewang95) [inline chat-widget/src/plugins/createChatTranscript.ts:214]: nit: trailing comma

File: chat-widget/src/plugins/createChatTranscript.ts [hunk 9]
```diff
@@ -210,9 +222,22 @@ class TranscriptHTMLBuilder {
                             return {
                                 ...activity,
                                 text: content,
-                                timestamp: created
+                                timestamp: created,
+                                channelData: { 
+                                    "webchat:sequence-id": webchatSequenceId
+                                },
```
Comment (charliewang95) [inline chat-widget/src/plugins/createChatTranscript.ts:228]: nit: trailing comma

File: chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/renderingmiddlewares/messageSequenceIdOverrideMiddleware.tsx [hunk 10]
```diff
@@ -0,0 +1,66 @@
+import { IWebChatAction } from "../../../interfaces/IWebChatAction";
```
Comment (charliewang95) [inline chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/renderingmiddlewares/messageSequenceIdOverrideMiddleware.tsx:1]: Can you please verify behavior after C2 disconnects (both close browser and network disconnect? C1 sending messages during C2 disconnect

Comment (elopezanaya) [reply chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/renderingmiddlewares/messageSequenceIdOverrideMiddleware.tsx:1]: image

No errors thrown when offline ,

Comment (charliewang95) [reply chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/renderingmiddlewares/messageSequenceIdOverrideMiddleware.tsx:1]: Could you verify after C2 disconnects, C1 sent messages order are also preserved after C2 reconnects?

**Review by charliewang95 — APPROVED** [review]

File: chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/renderingmiddlewares/messageSequenceIdOverrideMiddleware.tsx [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+    return Object.keys(channelData).some((key) => {
+        return (key === Constants.WebchatSequenceIdAttribute);
+    });
+};
+
+const overrideSequenceIdWithOriginalMessageId = (action: IWebChatAction): IWebChatAction => {
+
+    const originalMessageId = extractOriginalMessageId(action);
+    const channelData = action.payload.activity.channelData;
```
Comment (akhiloct1510) [inline chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/renderingmiddlewares/messageSequenceIdOverrideMiddleware.tsx:41]: We should add a null check on ChannelData. If for any reason or unknown scenario, this is not present, it will fail.

File: chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/renderingmiddlewares/messageSequenceIdOverrideMiddleware.tsx (hunk 10 above)
Comment (elopezanaya) [reply chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/renderingmiddlewares/messageSequenceIdOverrideMiddleware.tsx:1]: image
image

File: chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/renderingmiddlewares/messageSequenceIdOverrideMiddleware.tsx (hunk 11 above)
Comment (elopezanaya) [reply chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/renderingmiddlewares/messageSequenceIdOverrideMiddleware.tsx:40]: added an extra validation on isApplicable , an action with null channelData wont reach this block

**Review by akhiloct1510 — APPROVED** [review]
